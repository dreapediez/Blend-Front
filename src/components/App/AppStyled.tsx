import styled from "styled-components";

const AppStyled = styled.div`
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
`;
export default AppStyled;
