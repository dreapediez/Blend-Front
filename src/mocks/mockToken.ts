const mockToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6IjYzN2ZiNjgxODMyNjRlMDhkODkxNzdlZCIsInVzZXJuYW1lIjoiUGFjbyIsImlhdCI6MTY2OTMxNDQxOSwiZXhwIjoxNjY5NDg3MjE5fQ.zL_43PTvUh_yuOFS04Nd38_u5pNsAdSNWJM8byxRXu4";

export default mockToken;
