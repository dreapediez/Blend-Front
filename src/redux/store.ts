import { configureStore, ThunkAction, Action } from "@reduxjs/toolkit";
import { uiInitialState, uiReducer } from "./features/uiSlice/uiSlice";
import { userInitialState, userReducer } from "./features/userSlice/userSlice";

const preloadedState = {
  ui: uiInitialState,
  user: userInitialState,
};

export const store = ({ user, ui } = preloadedState) =>
  configureStore({
    preloadedState: {
      user,
      ui,
    },
    reducer: {
      user: userReducer,
      ui: uiReducer,
    },
  });

export const newStore = store();

export type AppDispatch = typeof newStore.dispatch;
export type RootState = ReturnType<typeof newStore.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
