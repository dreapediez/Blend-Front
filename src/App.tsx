import AppStyled from "./AppStyled";
import { Routes, Route, useLocation } from "react-router-dom";
import WelcomePage from "./pages/WelcomePage/WelcomePage";
import RegisterPage from "./pages/RegisterPage/RegisterPage";
import Header from "./components/Header/Header";
import Modal from "./components/Modal/Modal";
import { useAppSelector } from "./redux/hooks";
import LogInPage from "./pages/LogInPage/LogInPage";
import NotFoundPage from "./pages/NotFoundPage/NotFoundPage";
import Loading from "./components/Loading/Loading";
import CalendarPage from "./pages/ClaendarPage/CalendarPager";

const App = () => {
  const { pathname } = useLocation();
  const {
    isLoading,
    modal: { modalText, isError, showModal },
  } = useAppSelector((action) => action.ui);

  return (
    <AppStyled>
      {["/register", "/login", "/calendar"].includes(pathname) && <Header />}
      <Routes>
        <Route path="/" element={<WelcomePage />} />
        <Route path="/register" element={<RegisterPage />} />
        <Route path="/login" element={<LogInPage />} />
        <Route path="/calendar" element={<CalendarPage />} />
        <Route path="*" element={<NotFoundPage />} />
      </Routes>
      {isLoading && <Loading />}
      {showModal && (
        <Modal type={isError ? "error" : "success"} text={modalText} />
      )}
    </AppStyled>
  );
};

export default App;
