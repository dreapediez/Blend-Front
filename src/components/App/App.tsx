import AppStyled from "./AppStyled";
import { Routes, Route, useLocation } from "react-router-dom";
import { useAppSelector } from "../../redux/hooks";
import ProtectionRoutes from "../ProtectionRoutes/ProtectionRoutes";
import { useEffect, lazy, Suspense } from "react";
import useToken from "../../hooks/useToken/useToken";
import Modal from "../Modal/Modal";
import Loading from "../Loading/Loading";
import Header from "../Header/Header";
import PostFormPage from "../../pages/PostFormPage/PostFormPage";

const WelcomePageComponent = lazy(
  () => import("../../pages/WelcomePage/WelcomePage")
);
const RegisterPageComponent = lazy(
  () => import("../../pages/RegisterPage/RegisterPage")
);
const LogInPagePageComponent = lazy(
  () => import("../../pages/LogInPage/LogInPage")
);
const CalendarPageComponent = lazy(
  () => import("../../pages/CalendarPage/CalendarPage")
);
const PostsPageComponent = lazy(
  () => import("../../pages/PostDetailPage/PostDetailPage")
);
const NotFoundPageComponent = lazy(
  () => import("../../pages/NotFoundPage/NotFoundPage")
);

const App = () => {
  const { pathname } = useLocation();
  const { getToken } = useToken();
  const {
    isLoading,
    modal: { modalText, isError, showModal },
  } = useAppSelector((action) => action.ui);

  useEffect(() => {
    getToken();
  }, [getToken]);

  return (
    <AppStyled>
      {["/register", "/login", "/calendar", "/new-post"].includes(pathname) && (
        <Header />
      )}
      <Suspense fallback={<Loading />}>
        <Routes>
          <Route
            path="/"
            element={<ProtectionRoutes children={<WelcomePageComponent />} />}
          />
          <Route
            path="/register"
            element={<ProtectionRoutes children={<RegisterPageComponent />} />}
          />
          <Route
            path="/login"
            element={<ProtectionRoutes children={<LogInPagePageComponent />} />}
          />
          <Route
            path="/calendar"
            element={<ProtectionRoutes children={<CalendarPageComponent />} />}
          />
          <Route
            path="/detail-post/:idPost"
            element={<ProtectionRoutes children={<PostsPageComponent />} />}
          />
          <Route
            path="/new-post"
            element={<ProtectionRoutes children={<PostFormPage />} />}
          />
          <Route path="*" element={<NotFoundPageComponent />} />
        </Routes>
      </Suspense>
      {isLoading && <Loading />}
      {showModal && (
        <Modal type={isError ? "error" : "success"} text={modalText} />
      )}
    </AppStyled>
  );
};

export default App;
