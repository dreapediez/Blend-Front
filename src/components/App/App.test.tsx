import App from "./App";
import { MemoryRouter } from "react-router-dom";
import TestRenderer from "react-test-renderer";
import { Provider } from "react-redux";
import { act, render, screen, waitFor } from "@testing-library/react";
import {
  mockStore,
  mockStoreLoading,
  mockSuccessStore,
} from "../../mocks/makeWrapper";
import { newStore } from "../../redux/store";
import mockToken from "../../mocks/mockToken";

afterEach(() => {
  jest.clearAllMocks();
  localStorage.clear();
});

describe("Given an App component", () => {
  describe("When it's rendered with path '/'", () => {
    test("Then it should render 'not found page'", async () => {
      const expectedApp = TestRenderer.create(
        <Provider store={mockStoreLoading}>
          <MemoryRouter initialEntries={["/"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      expect(expectedApp).toMatchSnapshot();
    });
  });

  describe("When it's rendered with path '/login'", () => {
    test("Then it should render login page with their header component", async () => {
      const headerText = "Welcome to Blendcommunity!";

      // eslint-disable-next-line testing-library/no-unnecessary-act
      await act(() => {
        render(
          <Provider store={newStore}>
            <MemoryRouter initialEntries={["/login"]}>
              <App />
            </MemoryRouter>
          </Provider>
        );
      });

      await waitFor(() => {
        const expectedHeading = screen.queryByRole("heading", {
          name: headerText,
        });

        expect(expectedHeading).toBeInTheDocument();
      });
    });
  });

  describe("When it's rendered with path '/register'", () => {
    test("Then it should render register page with their header component", async () => {
      const headerText = "Join our community";

      render(
        <Provider store={newStore}>
          <MemoryRouter initialEntries={["/register"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      await waitFor(() => {
        const expectedHeading = screen.queryByRole("heading", {
          name: headerText,
        });

        expect(expectedHeading).toBeInTheDocument();
      });
    });
  });

  describe("When it's rendered with path '/calendar'", () => {
    test("Then it should render calendar page with their header component", async () => {
      const titleText = "Let’s open today’s window!";

      localStorage.setItem("token", mockToken);

      render(
        <Provider store={mockStoreLoading}>
          <MemoryRouter initialEntries={["/calendar"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      await waitFor(() => {
        const expectedHeading = screen.queryByRole("heading", {
          name: titleText,
        });

        expect(expectedHeading).toBeInTheDocument();
      });
    });
  });

  describe("When it's rendered with path '/posts'", () => {
    test("Then it should render posts page with their header component", async () => {
      const titleText = "Extra ingredients";

      localStorage.setItem("token", mockToken);

      // eslint-disable-next-line testing-library/no-unnecessary-act
      await act(() => {
        render(
          <Provider store={mockStoreLoading}>
            <MemoryRouter initialEntries={["/posts"]}>
              <App />
            </MemoryRouter>
          </Provider>
        );
      });

      await waitFor(() => {
        const expectedHeading = screen.queryAllByRole("heading", {
          name: titleText,
        });

        expect(expectedHeading[0]).toBeInTheDocument();
      });
    });
  });

  describe("When it's rendered with path '/404'", () => {
    test("Then it should render not found page with their header component", async () => {
      render(
        <Provider store={mockStoreLoading}>
          <MemoryRouter initialEntries={["/404"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      await waitFor(() => {
        const icon = screen.queryByTestId("404-icon");

        expect(icon).toBeInTheDocument();
      });
    });
  });

  describe("When it's render with Loading open", () => {
    test("Then it should render login page with their header component", async () => {
      const informationText = "loading";

      render(
        <Provider store={mockStoreLoading}>
          <MemoryRouter initialEntries={["/login"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      await waitFor(() => {
        const expectedModal = screen.queryByLabelText(informationText);

        expect(expectedModal).toBeInTheDocument();
      });
    });
  });

  describe("When it's render with an error", () => {
    test("Then it should show Modal Component with 'Something went wrong, please try again in a few minutes'", async () => {
      const informationText = "Error";
      render(
        <Provider store={mockStore}>
          <MemoryRouter initialEntries={["/"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      await waitFor(() => {
        const expectedModal = screen.queryByText(informationText);

        expect(expectedModal).toBeInTheDocument();
      });
    });
  });

  describe("When it's render with a success", () => {
    test("Then it should show Modal Component with 'Success'", async () => {
      const informationText = "Success";
      render(
        <Provider store={mockSuccessStore}>
          <MemoryRouter initialEntries={["/"]}>
            <App />
          </MemoryRouter>
        </Provider>
      );

      await waitFor(() => {
        const expectedModal = screen.queryByText(informationText);

        expect(expectedModal).toBeInTheDocument();
      });
    });
  });
});
