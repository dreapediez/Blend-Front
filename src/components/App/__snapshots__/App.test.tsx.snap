// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an App component When it's rendered with path '/' Then it should render 'not found page' 1`] = `
<div
  className="sc-bcXHqe eSuVKC"
>
  <div
    className="sc-jSUZER cDJIxu"
  >
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <span
      className="snowflake"
    />
    <header
      className="sc-gswNZR ipXOXK header"
    >
      <a
        href="/"
        onClick={[Function]}
      >
        <svg
          aria-label="A cup of tea with the text blend"
          className="welcome-page__logo"
          data-testid="logo-icon"
          width="360"
        >
          blend-logo.svg
        </svg>
      </a>
    </header>
    <div
      className="sc-jrcTuL fVAcvy"
    >
      <svg
        aria-label="Not page found"
        data-testid="404-icon"
        height="300"
        width="300"
      >
        404.svg
      </svg>
      <span
        className="form-footer__link"
      >
        Do you want some tea?
        <a
          className="form-footer__link-anchor"
          href="/calendar"
          onClick={[Function]}
        >
          Back home
        </a>
      </span>
    </div>
  </div>
</div>
`;
